### DISCUSSION TITLE
HTTP Client Span Names

### DISCUSSION THREAD

**Discussion by jamal** [open-telemetry/opentelemetry-java-instrumentation#2606] [Q&A]
Hi all! I wanted to try to start a quick conversation around the HTTP client span names provided by most instrumentations today. I know that we need to strike a balance between something that is specific enough, but provides low cardinality, and is not taxing performance. But, most instrumentations today seem to just be providing `HTTP {method}` as the span name.

I was wondering if it would make sense to include some additional information, such as the peer name, by default. I feel that this would at least allow for separation by external service which would provide far more valuable information. I noticed that in the base `HttpClientTracer` we are already using the `URI` from the implementation to set `net.peer.name`. Would there be any reason to not use this to generate a more unique span name, such as `HTTP {method} {hostname}`?

Thanks!

**Comment by trask** [comment]
hey @jamal, would you mind opening an issue for this in the specification repo? we would probably want to see [the spec guidance for http client span names](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#name) clarified to include this as a recommendation before making the change in this repo.

**Comment by genesiss** [comment]
Hi, here is a link to the specification: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span

There is an example table which says that GET is too broad of a name. Here is an example from Elastic APM, which shows why is this a problem when interpreting results: https://i.stack.imgur.com/ibxuN.png

**Reply by trask** [reply]
Hi @genesiss! check out https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#name, in particular

> HTTP client spans SHOULD be using conservative, low cardinality names formed from the available parameters of an HTTP request,
such as `"HTTP {METHOD_NAME}"`

**Reply by genesiss** [reply]
Hi, I see.. the link I posted above (same specification) somehow contradicts that. 

Here is an example what I see in Azure Insights if I use Open telemetry instrumentation and Azure Opentelemetry exporter:
![image](https://user-images.githubusercontent.com/273448/154633249-d0039315-c6e6-44fa-b472-b81c5dcd9441.png)

And here it is what I see when using Azure agent for instrumentation and export:
![azure_auto_end_to_end_tx](https://user-images.githubusercontent.com/273448/154633415-c4557098-d3ab-411b-9faf-a76b5fca0c63.PNG)

The second one is clearly more readable and informative than first one.

**Reply by laurit** [reply]
@genesiss These really don't contradict each other. As you have pointed out `HTTP GET` is a pretty bad name as it doesn't give any information about what the actual operation is. The problem here is that in general case we don't really know which part of the request is the operation name or which part has high cardinality in which case the spec requires us to use a guaranteed low cardinality name such as `HTTP GET`. To improve the situation we need to instrument higher level frameworks where it is possible to deduce a better operation name. For example with JAX-RS we can use `@Path` annotations and get nice operation names like `/get/{id}`.

**Reply by trask** [reply]
> The second one is clearly more readable and informative than first one.

This is due to different exporter mappings, not the underlying instrumentation.

**Reply by knutwannheden** [reply]
I also just stumbled across this. I did a quick search for a corresponding issue, but couldn't find anything except #6299, which is related.

Looking at https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/892fb8a38e4e3e8c7817c5836fe3902e9188fd3d/instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpSpanNameExtractor.java#L48-L53 it doesn't seem like any work has been done yet. Should I report an issue for this?

In contrast, gRPC appears to have dedicated client and server span name extractors as of #5244.

**Reply by trask** [reply]
(x-posted to #6526)
